## ТЗ
- Сравнить объекты в выгрузке из ЭБ с объектами в ЭР. Проверить, все ли объекты ЭР соответствуют объектам в ЭБ, в привязке к мероприятиям ЭР (результатам ФП в ЭБ).
- Сколько объектов и какого типа есть в ЭР сверх ЭБ

<i>*ОКС - объект капитального строительства</i>

In [1]:
#Иморт библиотек
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
#Отображение всех столбцов таблицы
pd.set_option('display.max_columns', None)

In [3]:
#Загружаемые файлы
vygr_eb = 'obj_EB_1211.xlsx' # - выгрузка из ГИИС "Электронный бюджет" за 12.11.2021
vygr_nsi_4856 = 'export_4856_1211.csv' # - выгрузка каталога 4856 (объекты) из НСИ за 12.11.2021
vygr_nsi_400 = 'export_400_1211.csv' # - выгрузка каталога 400 (мероприятия) из НСИ за 12.11.2021
er_976 = '976_lite.xlsx' # - файл с 976 мероприятиями Эталонного реестра

In [4]:
#Выгрузка из ЭБ по объектам в привязке с результатам ФП
eb_objects = pd.read_excel(vygr_eb)
eb_objects.head(3)

Код  национального проекта Наименование   национально проекта Код проекта  \
0                          A     Национальный проект "Культура"          A1   
1                          A     Национальный проект "Культура"          A1   
2                          A     Национальный проект "Культура"          A1   

                                Наименование проекта  \
0  Обеспечение качественно нового уровня развития...   
1  Обеспечение качественно нового уровня развития...   
2  Обеспечение качественно нового уровня развития...   

   Идентификатор результата ФП Код результата  \
0                         5917       A1.02.04   
1                         5917       A1.02.04   
2                         5917       A1.02.04   

                             Наименование результата  \
0  Построены (реконструированы) и (или) капитальн...   
1  Построены (реконструированы) и (или) капитальн...   
2  Построены (реконструированы) и (или) капитальн...   

                                      Тип результата  \
0  Строительство (реконструкция, техническое пере...   
1  Строительство (реконструкция, техническое пере...   
2  Строительство (реконструкция, техническое пере...   

  Срок реализации результата Реализуется регионом  ...  ФБ 2021 Всего 2022  \
0                 31.12.2024                   Да  ...      0.0        0.0   
1                 31.12.2024                   Да  ...      0.0        0.0   
2                 31.12.2024                   Да  ...  11000.0        0.0   

   ФБ 2022 Всего 2023 ФБ 2023 Всего 2024 ФБ 2024 Всего за весь срок  \
0      0.0        0.0     0.0        0.0     0.0                0.0   
1      0.0        0.0     0.0        0.0     0.0                0.0   
2      0.0        0.0     0.0        0.0     0.0                0.0   

  ФБ за весь срок Unnamed: 40  
0       300430.07         NaN  
1       313310.23         NaN  
2       110000.00         NaN  

[3 rows x 41 columns]

In [5]:
eb_objects.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3523 entries, 0 to 3522
Data columns (total 41 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   Код  национального проекта            3523 non-null   object 
 1   Наименование   национально проекта    3523 non-null   object 
 2   Код проекта                           3523 non-null   object 
 3   Наименование проекта                  3523 non-null   object 
 4   Идентификатор результата ФП           3523 non-null   int64  
 5   Код результата                        3523 non-null   object 
 6   Наименование результата               3523 non-null   object 
 7   Тип результата                        3523 non-null   object 
 8   Срок реализации результата            3523 non-null   object 
 9   Реализуется регионом                  3523 non-null   object 
 10  Идентификатор значения результата ФП  3523 non-null   int64  
 11  Срок реализации з

In [6]:
eb_objects['Идентификатор значения результата ФП'].head()

0    8216375245
1    8216375245
2    8216375245
3    8216375245
4    8216375245
Name: Идентификатор значения результата ФП, dtype: int64

In [7]:
eb_objects['Идентификатор результата ФП'].head()

0    5917
1    5917
2    5917
3    5917
4    5917
Name: Идентификатор результата ФП, dtype: int64

In [8]:
#Группировка объектов по резульататам ФП
eb_objects_by_results = eb_objects.groupby\
    (['Наименование результата', 'Идентификатор результата ФП', 'Идентификатор значения результата ФП'])\
    ['Наименование ОКС/объекта'].count()\
    .to_frame()\
    .reset_index()
eb_objects_by_results.rename(columns={'Наименование ОКС/объекта':'Кол-во ОКС/объектов'}, inplace=True)
eb_objects_by_results

Наименование результата  \
0    Будет проведена оценка необходимости принятия ...   
1    Введен в эксплуатацию модуль фабрикации и пуск...   
2    Введен в эксплуатацию учебно-тренировочный и и...   
3    Введены в эксплуатацию исследовательские станц...   
4    Внесение в законодательство Российской Федерац...   
..                                                 ...   
208  Уточнение порядка оказания финансовой поддержк...   
209  Финансовое обеспечение задачи для ввода в эксп...   
210  Финансовое обеспечение задачи для ввода в эксп...   
211  Финансовое обеспечение задачи для ввода в эксп...   
212  Финансовое обеспечение задачи для ввода в эксп...   

     Идентификатор результата ФП  Идентификатор значения результата ФП  \
0                           2551                            8217375254   
1                           2496                            8474163543   
2                           2497                            8474163547   
3                           7473                            8217171043   
4                           4891                            8497786263   
..                           ...                                   ...   
208                         2345                            8504633364   
209                         6173                            8216379186   
210                         6173                            8216379187   
211                         6173                            8216379188   
212                         6173                            8216418596   

     Кол-во ОКС/объектов  
0                      1  
1                      1  
2                      1  
3                      4  
4                      1  
..                   ...  
208                    1  
209                   10  
210                   14  
211                   19  
212                   22  

[213 rows x 4 columns]

In [9]:
eb_objects_by_results['Кол-во ОКС/объектов'].sum()

3523

<div class="alert alert-success">
В выгрузке из ЭБ у нас <b>213</b> результатов ФП и <b>3523</b> объекта.
</div>

In [10]:
eb_objects_by_results.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 213 entries, 0 to 212
Data columns (total 4 columns):
 #   Column                                Non-Null Count  Dtype 
---  ------                                --------------  ----- 
 0   Наименование результата               213 non-null    object
 1   Идентификатор результата ФП           213 non-null    int64 
 2   Идентификатор значения результата ФП  213 non-null    int64 
 3   Кол-во ОКС/объектов                   213 non-null    int64 
dtypes: int64(3), object(1)
memory usage: 6.8+ KB


In [11]:
#По выгрузке из ЭБ
print('Количество уникальных идентификаторов результатов ФП:', eb_objects_by_results['Идентификатор результата ФП'].nunique())
print('Количество уникальных идентификаторов значения результатов ФП:'\
      , eb_objects_by_results['Идентификатор значения результата ФП']\
      .nunique())

Количество уникальных идентификаторов результатов ФП: 132
Количество уникальных идентификаторов значения результатов ФП: 213


In [12]:
#Загрузка Эталонного реестра мероприятий 976
er_units_976 = pd.read_excel(er_976, sheet_name = '976')
er_units_976.head(3)

global_id system_object_id  \
0     201471       8498468734   
1     201472       8498454666   
2     201485       8497485987   

               Наименование мероприятия (результата)  \
0  На базе образовательных организаций высшего об...   
1  Педагогические работники и управленческие кадр...   
2  На базе общеобразовательных организаций создан...   

                                   Национальная цель  \
0  Возможности для самореализации и развития тала...   
1  Возможности для самореализации и развития тала...   
2  Возможности для самореализации и развития тала...   

                          Показатель достижения цели Национальный проект  \
0  Вхождение РФ в число 10-ти ведущих стран мира ...         Образование   
1  Вхождение РФ в число 10-ти ведущих стран мира ...         Образование   
2  Формирование эффективной системы выявления, по...         Образование   

  Государственная программа Федеральный проект  \
0      Развитие образования  Современная школа   
1      Развитие образования  Современная школа   
2      Развитие образования  Современная школа   

  Заместители Председателя Правительства  \
0            Голикова Татьяна Алексеевна   
1            Голикова Татьяна Алексеевна   
2            Голикова Татьяна Алексеевна   

                                     Организация  
0  Министерство просвещения Российской Федерации  
1  Министерство просвещения Российской Федерации  
2  Министерство просвещения Российской Федерации

In [13]:
er_units_976.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 976 entries, 0 to 975
Data columns (total 10 columns):
 #   Column                                  Non-Null Count  Dtype 
---  ------                                  --------------  ----- 
 0   global_id                               976 non-null    int64 
 1   system_object_id                        676 non-null    object
 2   Наименование мероприятия (результата)   976 non-null    object
 3   Национальная цель                       976 non-null    object
 4   Показатель достижения цели              976 non-null    object
 5   Национальный проект                     570 non-null    object
 6   Государственная программа               760 non-null    object
 7   Федеральный проект                      536 non-null    object
 8   Заместители Председателя Правительства  959 non-null    object
 9   Организация                             976 non-null    object
dtypes: int64(1), object(9)
memory usage: 76.4+ KB


In [14]:
#Список id 976 мероприятий Эталонного реестра из НСИ
list_976 = er_units_976['global_id'].to_list()

In [15]:
#Загрузка  реестра мероприятий из НСИ за 12.11.2021
nsi_units_1211 = pd.read_csv(vygr_nsi_400, delimiter = ';', encoding = 'ANSI', skiprows = 1)
nsi_units_1211.head(3)

global_id system_object_id  \
0     201471       8498468734   
1     201472       8498454666   
2     201485       8497485987   

               Наименование мероприятия (результата)  \
0  На базе образовательных организаций высшего об...   
1  Педагогические работники и управленческие кадр...   
2  На базе общеобразовательных организаций создан...   

   Статус заполнения карточки мероприятия  Позиция МинЭк  \
0                                     2.0            NaN   
1                                     2.0            NaN   
2                                     2.0            NaN   

                                   Комментарий МинЭк  \
0  Разблокировано в соответствии с резолюцией Гол...   
1  Разблокировано в соответствии с резолюцией Гол...   
2  Разблокировано в соответствии с резолюцией Гол...   

                                      Комментарий АЦ  Позиция ведомства  \
0  1. Реализация в Регионах - отсутствует. При эт...                NaN   
1                                        Согласовано                NaN   
2  1. Плановое значение результата в ежемесячном ...                NaN   

   Тип мероприятия (результата) Тип мероприятия (результата) (временно)  ...  \
0                            71                                     NaN  ...   
1                            61                                     NaN  ...   
2                            68      Приобретение товаров, работ, услуг  ...   

   Статус  Удалить. Техническая правка  Причина удаления    Состояние подписи  \
0       5                          NaN               NaN  Запись не подписана   
1       5                          NaN               NaN  Запись не подписана   
2       5                          NaN               NaN  Запись не подписана   

   Тип последнего изменения  Дата последнего изменения      Дата подписания  \
0                 изменение        09.11.2021 15:50:03  25.05.2021 12:04:15   
1                 изменение        09.11.2021 15:52:02  14.05.2021 19:34:20   
2                 изменение        09.11.2021 16:06:12  14.05.2021 19:34:20   

  Пользователь/система производившие изменения        Дата создания  \
0                Кветинская Алеся Владимировна  01.04.2021 00:43:22   
1                Кветинская Алеся Владимировна  01.04.2021 00:43:06   
2                Кветинская Алеся Владимировна  01.04.2021 00:43:37   

  Статус объекта  
0        активна  
1        активна  
2        активна  

[3 rows x 66 columns]

In [16]:
#Количество мероприятиц в выгрузке НСИ
len(nsi_units_1211)

1104

In [17]:
#Функция определения состоит ли id мероприятия в Эталонном реестре
def id_in_er(data):
    if data in list_976:
        data = 1
    else:
        data = 0
    return data

In [18]:
#Применение функции и создание нового столбца
nsi_units_1211['ЭР'] = nsi_units_1211['global_id'].apply(id_in_er)

In [19]:
#Отфильтруем из выгрузки НСИ за 12.11.2021 мероприятия, содержащиеся в Эталонном реестре
nsi_er_units_1211 = nsi_units_1211[nsi_units_1211['ЭР'] == 1]
len(nsi_er_units_1211)

972

<div class="alert alert-success">
Таким образом получаем, что в НСИ по состоянию на 12.11.2021 загружено <b>972</b> мероприятия Эталонного реестра (из 976).
    </div>

In [20]:
nsi_er_units_1211.head(3)

global_id system_object_id  \
0     201471       8498468734   
1     201472       8498454666   
2     201485       8497485987   

               Наименование мероприятия (результата)  \
0  На базе образовательных организаций высшего об...   
1  Педагогические работники и управленческие кадр...   
2  На базе общеобразовательных организаций создан...   

   Статус заполнения карточки мероприятия  Позиция МинЭк  \
0                                     2.0            NaN   
1                                     2.0            NaN   
2                                     2.0            NaN   

                                   Комментарий МинЭк  \
0  Разблокировано в соответствии с резолюцией Гол...   
1  Разблокировано в соответствии с резолюцией Гол...   
2  Разблокировано в соответствии с резолюцией Гол...   

                                      Комментарий АЦ  Позиция ведомства  \
0  1. Реализация в Регионах - отсутствует. При эт...                NaN   
1                                        Согласовано                NaN   
2  1. Плановое значение результата в ежемесячном ...                NaN   

   Тип мероприятия (результата) Тип мероприятия (результата) (временно)  ...  \
0                            71                                     NaN  ...   
1                            61                                     NaN  ...   
2                            68      Приобретение товаров, работ, услуг  ...   

   Удалить. Техническая правка  Причина удаления    Состояние подписи  \
0                          NaN               NaN  Запись не подписана   
1                          NaN               NaN  Запись не подписана   
2                          NaN               NaN  Запись не подписана   

   Тип последнего изменения  Дата последнего изменения      Дата подписания  \
0                 изменение        09.11.2021 15:50:03  25.05.2021 12:04:15   
1                 изменение        09.11.2021 15:52:02  14.05.2021 19:34:20   
2                 изменение        09.11.2021 16:06:12  14.05.2021 19:34:20   

  Пользователь/система производившие изменения        Дата создания  \
0                Кветинская Алеся Владимировна  01.04.2021 00:43:22   
1                Кветинская Алеся Владимировна  01.04.2021 00:43:06   
2                Кветинская Алеся Владимировна  01.04.2021 00:43:37   

  Статус объекта ЭР  
0        активна  1  
1        активна  1  
2        активна  1  

[3 rows x 67 columns]

In [21]:
#Создадим сокращенную таблицу с нужными полями
clean_nsi_er_units_1211 = nsi_er_units_1211[['global_id', 'system_object_id', 'Наименование мероприятия (результата)', \
                                             'Тип мероприятия (результата)', \
                                             'Код из Электронного бюджета']]
clean_nsi_er_units_1211

global_id system_object_id  \
0       201471       8498468734   
1       201472       8498454666   
2       201485       8497485987   
3       201497       8498443267   
4       201501       8497503134   
..         ...              ...   
967     247628              NaN   
968     247630              NaN   
969     250736              NaN   
970     251522              NaN   
971     252241              NaN   

                 Наименование мероприятия (результата)  \
0    На базе образовательных организаций высшего об...   
1    Педагогические работники и управленческие кадр...   
2    На базе общеобразовательных организаций создан...   
3    Сформирована и функционирует единая федеральна...   
4    Проведено повышение квалификации штатных педаг...   
..                                                 ...   
967  Созданы инвестиционные комитеты и агенства раз...   
968  Субъектам МСП предоставлено имущество (в аренд...   
969  Субъекты МСП обеспечены обучающими мероприятия...   
970  Принят Федеральный закон "О внесении изменений...   
971  Увеличен объем международных почтовых отправле...   

     Тип мероприятия (результата) Код из Электронного бюджета  
0                              71                  8498468734  
1                              61                  8498454666  
2                              68                  8497485987  
3                              71                  8498443267  
4                              61                  8497503134  
..                            ...                         ...  
967                            66                         NaN  
968                            71                  8498634064  
969                            61                  8498668170  
970                            62                         NaN  
971                            84                  8667262825  

[972 rows x 5 columns]

In [22]:
#Загрузка  реестра объектов (каталог 4856) из НСИ за 12.11.2021
nsi_objects_1211 = pd.read_csv(vygr_nsi_4856, delimiter = ';', encoding = 'ANSI', skiprows = 1)
nsi_objects_1211.head(3)

global_id  system_object_id  global_object_id system_global_object_id  \
0        577               NaN            201616                     340   
1        578               NaN            201616                     340   
2        855               NaN            201616                     340   

   Субъект РФ                               Наименование объекта  \
0         6.0  Школа на 720 мест г. Назрани. (кадастровым ном...   
1         6.0  Школа на 720 мест в с.п. Троицкое в 2021 году ...   
2         6.0  Школа на 720 мест в г. Назрани. (кадастровым н...   

        Местонахождение объекта  Ед. измерения  Плановое значение  \
0  г. Назрань, ул. Дагестанская             NaN               NaN   
1                 с.п. Троицкое             NaN               NaN   
2  г. Назрань, ул. Тумгоева, 10             NaN               NaN   

  Фактическое значение                                  Состояние подписи  \
0                  NaN  Запись подписана 14.05.2021 19:43:20 CN=МИНИСТ...   
1                  NaN  Запись подписана 14.05.2021 19:43:20 CN=МИНИСТ...   
2                  NaN  Запись подписана 14.05.2021 19:43:20 CN=МИНИСТ...   

  Статус объекта  
0        активна  
1        активна  
2        активна

In [23]:
#Отфильтруем из выгрузки объектов НСИ за 12.11.2021 объекты, привязанные к мероприятия, содержащиеся в Эталонном реестре
nsi_objects_1211['ЭР'] = nsi_objects_1211['global_object_id'].apply(id_in_er)
nsi_er_objects_1211 = nsi_objects_1211[nsi_objects_1211['ЭР'] == 1]
nsi_er_objects_1211.head(3)

global_id  system_object_id  global_object_id system_global_object_id  \
0        577               NaN            201616                     340   
1        578               NaN            201616                     340   
2        855               NaN            201616                     340   

   Субъект РФ                               Наименование объекта  \
0         6.0  Школа на 720 мест г. Назрани. (кадастровым ном...   
1         6.0  Школа на 720 мест в с.п. Троицкое в 2021 году ...   
2         6.0  Школа на 720 мест в г. Назрани. (кадастровым н...   

        Местонахождение объекта  Ед. измерения  Плановое значение  \
0  г. Назрань, ул. Дагестанская             NaN               NaN   
1                 с.п. Троицкое             NaN               NaN   
2  г. Назрань, ул. Тумгоева, 10             NaN               NaN   

  Фактическое значение                                  Состояние подписи  \
0                  NaN  Запись подписана 14.05.2021 19:43:20 CN=МИНИСТ...   
1                  NaN  Запись подписана 14.05.2021 19:43:20 CN=МИНИСТ...   
2                  NaN  Запись подписана 14.05.2021 19:43:20 CN=МИНИСТ...   

  Статус объекта  ЭР  
0        активна   1  
1        активна   1  
2        активна   1

In [24]:
nsi_er_objects_1211.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 125392 entries, 0 to 126846
Data columns (total 13 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   global_id                125392 non-null  int64  
 1   system_object_id         37057 non-null   float64
 2   global_object_id         125392 non-null  int64  
 3   system_global_object_id  45896 non-null   object 
 4   Субъект РФ               125391 non-null  float64
 5   Наименование объекта     125379 non-null  object 
 6   Местонахождение объекта  125012 non-null  object 
 7   Ед. измерения            123093 non-null  float64
 8   Плановое значение        123212 non-null  object 
 9   Фактическое значение     31885 non-null   object 
 10  Состояние подписи        125392 non-null  object 
 11  Статус объекта           125392 non-null  object 
 12  ЭР                       125392 non-null  int64  
dtypes: float64(3), int64(3), object(7)
memory usage: 13.4+ MB


In [25]:
print('Количество объектов в выгрузке НСИ за 12.11.2021 привязанных к мероприятиям ЭР:', len(nsi_er_objects_1211))

Количество объектов в выгрузке НСИ за 12.11.2021 привязанных к мероприятиям ЭР: 125392


<div class="alert alert-success">
<b>Сравнение выгрузок</b> (по состоянию на 12.11.2021):

- В выгрузке из ГИИС "ЭБ" содержится <b>3523</b> объекта
- В выгрузке из НСИ в привязке к мероприятиям Эталонного реестра содержится <b>125392</b> объекта

In [26]:
#Группировка объектов по мероприятиям
nsi_objects_by_units = nsi_er_objects_1211.groupby('global_object_id')['global_id'].count()\
    .to_frame()\
    .reset_index()
nsi_objects_by_units.rename(columns={'global_id':'Кол-во объектов'}, inplace=True)
nsi_objects_by_units

global_object_id  Кол-во объектов
0              201485               48
1              201535             2125
2              201581              229
3              201584              139
4              201588               55
..                ...              ...
171            244091                1
172            244092                1
173            244093                1
174            244094                1
175            244559                1

[176 rows x 2 columns]

In [27]:
#Проверка количества объектов
nsi_objects_by_units['Кол-во объектов'].sum()

125392

In [28]:
#Проверка количества уникальных id мероприятий
nsi_er_objects_1211['global_object_id'].nunique()

176

In [29]:
#Соединим таблицы чтобы получить реквизиты мероприятия
nsi_objects_by_units = clean_nsi_er_units_1211.merge(nsi_objects_by_units, how = 'left'\
                                                     , left_on = 'global_id', right_on = 'global_object_id')
nsi_objects_by_units

global_id system_object_id  \
0       201471       8498468734   
1       201472       8498454666   
2       201485       8497485987   
3       201497       8498443267   
4       201501       8497503134   
..         ...              ...   
967     247628              NaN   
968     247630              NaN   
969     250736              NaN   
970     251522              NaN   
971     252241              NaN   

                 Наименование мероприятия (результата)  \
0    На базе образовательных организаций высшего об...   
1    Педагогические работники и управленческие кадр...   
2    На базе общеобразовательных организаций создан...   
3    Сформирована и функционирует единая федеральна...   
4    Проведено повышение квалификации штатных педаг...   
..                                                 ...   
967  Созданы инвестиционные комитеты и агенства раз...   
968  Субъектам МСП предоставлено имущество (в аренд...   
969  Субъекты МСП обеспечены обучающими мероприятия...   
970  Принят Федеральный закон "О внесении изменений...   
971  Увеличен объем международных почтовых отправле...   

     Тип мероприятия (результата) Код из Электронного бюджета  \
0                              71                  8498468734   
1                              61                  8498454666   
2                              68                  8497485987   
3                              71                  8498443267   
4                              61                  8497503134   
..                            ...                         ...   
967                            66                         NaN   
968                            71                  8498634064   
969                            61                  8498668170   
970                            62                         NaN   
971                            84                  8667262825   

     global_object_id  Кол-во объектов  
0                 NaN              NaN  
1                 NaN              NaN  
2            201485.0             48.0  
3                 NaN              NaN  
4                 NaN              NaN  
..                ...              ...  
967               NaN              NaN  
968               NaN              NaN  
969               NaN              NaN  
970               NaN              NaN  
971               NaN              NaN  

[972 rows x 7 columns]

In [30]:
#Загрузка справочника типов мероприятий
spr_unit_type = pd.read_excel('spr_unit_type.xlsx')
spr_unit_type

Идентификатор                                       Наименование
0              57  Создание (развитие) информационно-телекоммуник...
1              58  Создание (реорганизация) организации (структур...
2              59               Привлечение квалифицированных кадров
3              60   Создание Российской промышленной зоны за рубежом
4              61             Проведение образовательных мероприятий
5              62   Принятие нормативного правового (правового) акта
6              63  Обеспечение реализации федерального (ведомстве...
7              64                    Проведение массовых мероприятий
8              65  Строительство (реконструкция, техническое пере...
9              66                              Утверждение документа
10             67                         Благоустройство территории
11             68                 Приобретение товаров, работ, услуг
12             69  Проведение научно-исследовательских (опытно-ко...
13             70  Проведение информационно-коммуникационной камп...
14             71                  Оказание услуг (выполнение работ)
15             72       Налоговые льготы (стимулирующие, социальные)
16             73  Внедрение новой технологии (в том числе новой ...
17             74              Ремонт объектов недвижимого имущества
18             75                Обслуживание государственного долга
19             76                          Меры социальной поддержки
20             77  Меры поддержки субъектов или отдельных террито...
21             78  Обеспечение органов государственной власти (в ...
22             79                                Субвенции и дотации
23             80     Оказание государственных (муниципальных) услуг
24             81                   Ликвидация чрезвычайной ситуации
25             82  Финансирование и исполнение госзаданий (субсид...
26             83  Исполнение международных обязательств (имеющие...
27             84  Меры господдержки (в т.ч. субсидии юридическим...

In [31]:
#Подгрузим справочник типов мероприятий
nsi_objects_by_units = nsi_objects_by_units.merge(spr_unit_type, how = 'left', left_on = 'Тип мероприятия (результата)', \
                           right_on = 'Идентификатор')
nsi_objects_by_units

global_id system_object_id  \
0       201471       8498468734   
1       201472       8498454666   
2       201485       8497485987   
3       201497       8498443267   
4       201501       8497503134   
..         ...              ...   
967     247628              NaN   
968     247630              NaN   
969     250736              NaN   
970     251522              NaN   
971     252241              NaN   

                 Наименование мероприятия (результата)  \
0    На базе образовательных организаций высшего об...   
1    Педагогические работники и управленческие кадр...   
2    На базе общеобразовательных организаций создан...   
3    Сформирована и функционирует единая федеральна...   
4    Проведено повышение квалификации штатных педаг...   
..                                                 ...   
967  Созданы инвестиционные комитеты и агенства раз...   
968  Субъектам МСП предоставлено имущество (в аренд...   
969  Субъекты МСП обеспечены обучающими мероприятия...   
970  Принят Федеральный закон "О внесении изменений...   
971  Увеличен объем международных почтовых отправле...   

     Тип мероприятия (результата) Код из Электронного бюджета  \
0                              71                  8498468734   
1                              61                  8498454666   
2                              68                  8497485987   
3                              71                  8498443267   
4                              61                  8497503134   
..                            ...                         ...   
967                            66                         NaN   
968                            71                  8498634064   
969                            61                  8498668170   
970                            62                         NaN   
971                            84                  8667262825   

     global_object_id  Кол-во объектов  Идентификатор  \
0                 NaN              NaN             71   
1                 NaN              NaN             61   
2            201485.0             48.0             68   
3                 NaN              NaN             71   
4                 NaN              NaN             61   
..                ...              ...            ...   
967               NaN              NaN             66   
968               NaN              NaN             71   
969               NaN              NaN             61   
970               NaN              NaN             62   
971               NaN              NaN             84   

                                          Наименование  
0                    Оказание услуг (выполнение работ)  
1               Проведение образовательных мероприятий  
2                   Приобретение товаров, работ, услуг  
3                    Оказание услуг (выполнение работ)  
4               Проведение образовательных мероприятий  
..                                                 ...  
967                              Утверждение документа  
968                  Оказание услуг (выполнение работ)  
969             Проведение образовательных мероприятий  
970   Принятие нормативного правового (правового) акта  
971  Меры господдержки (в т.ч. субсидии юридическим...  

[972 rows x 9 columns]

In [32]:
#Количество мероприятий ЭР с объектами
len(nsi_objects_by_units[nsi_objects_by_units['Кол-во объектов'].notna()])

176

In [33]:
#Количество объектов в этих мероприятиях
nsi_objects_by_units[nsi_objects_by_units['Кол-во объектов'].notna()]['Кол-во объектов'].sum()

125392.0

<div class="alert alert-success">
В НСИ по состоянию на 12.11.2021 из 972 мероприятий Эталонного реестра <b>176</b> мероприятий содержат информацию об объектах. Всего в указанных 176 мероприятиях содержится информация о <b>125392</b> объектах.
</div>

In [34]:
eb_objects_by_results

Наименование результата  \
0    Будет проведена оценка необходимости принятия ...   
1    Введен в эксплуатацию модуль фабрикации и пуск...   
2    Введен в эксплуатацию учебно-тренировочный и и...   
3    Введены в эксплуатацию исследовательские станц...   
4    Внесение в законодательство Российской Федерац...   
..                                                 ...   
208  Уточнение порядка оказания финансовой поддержк...   
209  Финансовое обеспечение задачи для ввода в эксп...   
210  Финансовое обеспечение задачи для ввода в эксп...   
211  Финансовое обеспечение задачи для ввода в эксп...   
212  Финансовое обеспечение задачи для ввода в эксп...   

     Идентификатор результата ФП  Идентификатор значения результата ФП  \
0                           2551                            8217375254   
1                           2496                            8474163543   
2                           2497                            8474163547   
3                           7473                            8217171043   
4                           4891                            8497786263   
..                           ...                                   ...   
208                         2345                            8504633364   
209                         6173                            8216379186   
210                         6173                            8216379187   
211                         6173                            8216379188   
212                         6173                            8216418596   

     Кол-во ОКС/объектов  
0                      1  
1                      1  
2                      1  
3                      4  
4                      1  
..                   ...  
208                    1  
209                   10  
210                   14  
211                   19  
212                   22  

[213 rows x 4 columns]

In [35]:
#Посчитаем кол-во уникальных идентификаторов результата ФП
eb_objects_by_results['Идентификатор результата ФП'].nunique()

132

In [36]:
#Посчитаем кол-во уникальных идентификаторов значения результата ФП
eb_objects_by_results['Идентификатор значения результата ФП'].nunique()

213

<div class="alert alert-success">
Очевидно, что к результату может быть привязано несколько значений результата (со своими id)
</div>

In [37]:
#Сформируем списки id результата ФП и id значения результата ФП
id_results_eb = eb_objects_by_results['Идентификатор результата ФП'].unique().tolist()
id_values_results_eb = eb_objects_by_results['Идентификатор значения результата ФП'].unique().tolist()

In [38]:
def id_in_spisok(data):
    if data in spisok:
        data = 1
    else:
        data = 0
    return data

In [39]:
#Проверим есть ли совпадения в мероприятиях НСИ с ЭБ по id "Идентификатор результата ФП"
spisok = id_results_eb
nsi_objects_by_units['match'] = nsi_objects_by_units['system_object_id'].apply(id_in_spisok)
nsi_objects_by_units[nsi_objects_by_units['match'] == 1]

Empty DataFrame
Columns: [global_id, system_object_id, Наименование мероприятия (результата), Тип мероприятия (результата), Код из Электронного бюджета, global_object_id, Кол-во объектов, Идентификатор, Наименование, match]
Index: []

In [40]:
nsi_objects_by_units['match'] = nsi_objects_by_units['Код из Электронного бюджета'].apply(id_in_spisok)
nsi_objects_by_units[nsi_objects_by_units['match'] == 1]

Empty DataFrame
Columns: [global_id, system_object_id, Наименование мероприятия (результата), Тип мероприятия (результата), Код из Электронного бюджета, global_object_id, Кол-во объектов, Идентификатор, Наименование, match]
Index: []

<div class="alert alert-success">
Таким образом, мы выяснили что "Идентификатор результата ФП" не присутствует в полях выгрузки из НСИ.
</div>

In [41]:
#Проверим есть ли совпадения в мероприятиях НСИ с ЭБ по id "Идентификатор значения результата ФП"
spisok = id_values_results_eb
nsi_objects_by_units['match'] = nsi_objects_by_units['system_object_id'].apply(id_in_spisok)
nsi_objects_by_units[nsi_objects_by_units['match'] == 1]

Empty DataFrame
Columns: [global_id, system_object_id, Наименование мероприятия (результата), Тип мероприятия (результата), Код из Электронного бюджета, global_object_id, Кол-во объектов, Идентификатор, Наименование, match]
Index: []

In [42]:
nsi_objects_by_units['match'] = nsi_objects_by_units['Код из Электронного бюджета'].apply(id_in_spisok)
nsi_objects_by_units[nsi_objects_by_units['match'] == 1]

Empty DataFrame
Columns: [global_id, system_object_id, Наименование мероприятия (результата), Тип мероприятия (результата), Код из Электронного бюджета, global_object_id, Кол-во объектов, Идентификатор, Наименование, match]
Index: []

<div class="alert alert-success">
Также и "Идентификатор значения результата" не находится в выгрузке НСИ.
</div>

In [43]:
#Проверка функции
spisok = id_values_results_eb
eb_objects_by_results['match'] = eb_objects_by_results['Идентификатор значения результата ФП'].apply(id_in_spisok)
len(eb_objects_by_results[eb_objects_by_results['match'] == 1])

213

<div class="alert alert-success">
Функция работает корректно. В выгрузке из ЭБ все находит.
</div>

In [44]:
nsi_objects_by_units.head(3)

global_id system_object_id  \
0     201471       8498468734   
1     201472       8498454666   
2     201485       8497485987   

               Наименование мероприятия (результата)  \
0  На базе образовательных организаций высшего об...   
1  Педагогические работники и управленческие кадр...   
2  На базе общеобразовательных организаций создан...   

   Тип мероприятия (результата) Код из Электронного бюджета  global_object_id  \
0                            71                  8498468734               NaN   
1                            61                  8498454666               NaN   
2                            68                  8497485987          201485.0   

   Кол-во объектов  Идентификатор                            Наименование  \
0              NaN             71       Оказание услуг (выполнение работ)   
1              NaN             61  Проведение образовательных мероприятий   
2             48.0             68      Приобретение товаров, работ, услуг   

   match  
0      0  
1      0  
2      0

In [45]:
#Cгруппируем кол-во объектов в привязке к результатам ФП в выгрузке из ЭБ
merged_eb_objects_by_results = eb_objects_by_results.groupby\
    ('Наименование результата')['Кол-во ОКС/объектов'].sum().to_frame().reset_index()
merged_eb_objects_by_results

Наименование результата  Кол-во ОКС/объектов
0    Будет проведена оценка необходимости принятия ...                    1
1    Введен в эксплуатацию модуль фабрикации и пуск...                    1
2    Введен в эксплуатацию учебно-тренировочный и и...                    1
3    Введены в эксплуатацию исследовательские станц...                    4
4    Внесение в законодательство Российской Федерац...                    1
..                                                 ...                  ...
127  Утвержден перечень субъектов Российской Федера...                    4
128  Утверждены критерии признания земельных участк...                    1
129  Утверждены нормативные правовые акты по регист...                    1
130  Уточнение порядка оказания финансовой поддержк...                    1
131  Финансовое обеспечение задачи для ввода в эксп...                   65

[132 rows x 2 columns]

In [46]:
merged_eb_objects_by_results['Кол-во ОКС/объектов'].sum()

3523

<div class="alert alert-success">
В выгрузке из ГИИС "Электронный бюджет" <b>3523</b> объекта привязаны к <b>132</b> результатам федеральных проектов.

<div class="alert alert-success">
Так как мы выяснили, что ни по каким id у нас совпадений в выгрузках нет, придется соединять полученные таблицы по по наименованию мероприятия (результата).
</div>

In [47]:
#Соединим выгрузки НСИ и ЭБ (по названию мероприятия)
match = nsi_objects_by_units.merge(merged_eb_objects_by_results, how = 'inner', \
                                   left_on = 'Наименование мероприятия (результата)',\
                           right_on = 'Наименование результата')
match

global_id system_object_id  \
0      201581              710   
1      201615       8216400552   
2      201616              340   
3      201626       8216397653   
4      201824            21115   
5      202574       8216349767   
6      204923       8499609862   
7      204992       8495457932   
8      205004            20923   
9      205033            20014   
10     205124       8498038763   
11     205303       8216385548   
12     205817            21868   
13     205894       8216397648   
14     205897       8217077149   
15     205927       8216598824   
16     207569       8513960443   
17     207586       8474169452   
18     207588       8513817854   
19     207614       8474163536   
20     207621       8216973329   
21     207648       8217682350   
22     207651       8474169446   
23     243516              NaN   

                Наименование мероприятия (результата)  \
0   Создано новых мест в общеобразовательных орган...   
1   Создано новых мест в общеобразовательных орган...   
2   Создано новых мест в общеобразовательных орган...   
3   Реализованы мероприятия по модернизации инфрас...   
4   Построены два новых современных научно-исследо...   
5   Разработаны и приняты нормативные правовые акт...   
6   Утверждены критерии признания земельных участк...   
7   Реконструированы и (или) капитально отремонтир...   
8   Проведена реновация федеральных организаций от...   
9   Реализованы проекты по развитию территорий, ра...   
10  Утверждены нормативные правовые акты по регист...   
11  Построены и введены в эксплуатацию объекты спо...   
12  Финансовое обеспечение задачи для ввода в эксп...   
13  Создание и эксплуатация подмосковного образова...   
14  Создание и эксплуатация образовательного центр...   
15  Создан информационно-вычислительный центр ФГБУ...   
16  Завершена реконструкция и введен в эксплуатаци...   
17  Завершено техническое перевооружение и введен ...   
18  Завершена реконструкция и введен в эксплуатаци...   
19  Завершено техническое перевооружение и введен ...   
20  Осуществлена реконструкция объектов в аэропорт...   
21  Завершено строительство третьего и начато стро...   
22  Завершено техническое перевооружение и введен ...   
23  Созданы дополнительные места для детей в возра...   

    Тип мероприятия (результата) Код из Электронного бюджета  \
0                             65                         710   
1                             65                  8216400552   
2                             65                         340   
3                             65                  8216397653   
4                             65                       21115   
5                             62                  8216349767   
6                             62                  8499609862   
7                             77                  8495457932   
8                             65                       20923   
9                             66                       20014   
10                            62                  8498038763   
11                            65                  8216385548   
12                            65                       21868   
13                            65                  8216397648   
14                            65                  8217077149   
15                            58                  8216598824   
16                            65                  8513960443   
17                            65                  8474169452   
18                            65                  8513817854   
19                            65                  8474163536   
20                            65                  8216973329   
21                            65                  8217682350   
22                            65                  8474169446   
23                            65                          75   

    global_object_id  Кол-во объектов  Идентификатор  \
0           201581.0            229.0             65   
1           20

In [48]:
#Проверка на дубликаты
match['global_id'].nunique()

24

In [49]:
#Длина таблицы
len(match)

24

In [50]:
#Приведем таблицу с результатом к удобному виду
match.drop(['system_object_id', 'global_object_id', 'match', 'Наименование результата'], axis = 1, inplace = True)

match = match[['global_id', 'Код из Электронного бюджета', 'Наименование мероприятия (результата)', 'Кол-во объектов'\
      , 'Идентификатор', 'Наименование', 'Кол-во ОКС/объектов']].copy()

match.rename(columns = {'Кол-во объектов':'Кол-во_объектов_ЭР','Идентификатор':'Тип мероприятия', \
                        'Наименование':'Наименование типа мероприятия', 'Кол-во ОКС/объектов':'Кол-во_объектов_ЭБ'},\
             inplace = True)

match['Кол-во_объектов_ЭР'].fillna(value = 0, inplace = True)

match['Кол-во_объектов_ЭР'] = match['Кол-во_объектов_ЭР'].astype('int32')

In [51]:
#Добавим столбец с разницей
match['Разница ЭР - ЭБ'] = match['Кол-во_объектов_ЭР'] - match['Кол-во_объектов_ЭБ']

In [52]:
match

global_id Код из Электронного бюджета  \
0      201581                         710   
1      201615                  8216400552   
2      201616                         340   
3      201626                  8216397653   
4      201824                       21115   
5      202574                  8216349767   
6      204923                  8499609862   
7      204992                  8495457932   
8      205004                       20923   
9      205033                       20014   
10     205124                  8498038763   
11     205303                  8216385548   
12     205817                       21868   
13     205894                  8216397648   
14     205897                  8217077149   
15     205927                  8216598824   
16     207569                  8513960443   
17     207586                  8474169452   
18     207588                  8513817854   
19     207614                  8474163536   
20     207621                  8216973329   
21     207648                  8217682350   
22     207651                  8474169446   
23     243516                          75   

                Наименование мероприятия (результата)  Кол-во_объектов_ЭР  \
0   Создано новых мест в общеобразовательных орган...                 229   
1   Создано новых мест в общеобразовательных орган...                 215   
2   Создано новых мест в общеобразовательных орган...                  88   
3   Реализованы мероприятия по модернизации инфрас...                  76   
4   Построены два новых современных научно-исследо...                   1   
5   Разработаны и приняты нормативные правовые акт...                   0   
6   Утверждены критерии признания земельных участк...                   0   
7   Реконструированы и (или) капитально отремонтир...                  20   
8   Проведена реновация федеральных организаций от...                   0   
9   Реализованы проекты по развитию территорий, ра...                 470   
10  Утверждены нормативные правовые акты по регист...                   0   
11  Построены и введены в эксплуатацию объекты спо...                  68   
12  Финансовое обеспечение задачи для ввода в эксп...                  10   
13  Создание и эксплуатация подмосковного образова...                   1   
14  Создание и эксплуатация образовательного центр...                   1   
15  Создан информационно-вычислительный центр ФГБУ...                   0   
16  Завершена реконструкция и введен в эксплуатаци...                   1   
17  Завершено техническое перевооружение и введен ...                   1   
18  Завершена реконструкция и введен в эксплуатаци...                   1   
19  Завершено техническое перевооружение и введен ...                   1   
20  Осуществлена реконструкция объектов в аэропорт...                  53   
21  Завершено строительство третьего и начато стро...                   2   
22  Завершено техническое перевооружение и введен ...                   1   
23  Созданы дополнительные места для детей в возра...                   0   

    Тип мероприятия                      Наименование типа мероприятия  \
0                65  Строительство (реконструкция, техническое пере...   
1                65  Строительство (реконструкция, техническое пере...   
2                65  Строительство (реконструкция, техническое пере...   
3                65  Строительство (реконструкция, техническое пере...   
4                65  Строительство (реконструкция, техническое пере...   
5                62   Принятие нормативного правового (правового) акта   
6                62   Принятие нормативного правового (правового) акта   
7                77  Меры поддержки субъектов или отдельных террито...   
8                65  Строительство (реконструкция, техническое пере...   
9                66                              Утверждение документа   
10               62   Принятие нормативного правового (правового) акта   
11               65  Строительство (реконструкция, техническое п

<div class="alert alert-success">
Получилось пересечение по 24 мероприятиям.
</div>

In [53]:
#Выгрузка результата в Excel
#match.to_excel('output_7.xlsx')

<div class="alert alert-success">
<H1>Выводы</H1>
    
В НСИ по состоянию на 12.11.2021 загружено <b>972</b> мероприятия Эталонного реестра (из 976).

Сравнение выгрузок (по состоянию на 12.11.2021):
В выгрузке из ГИИС "ЭБ" содержится <b>3523</b> объекта
В выгрузке из НСИ в привязке к мероприятиям Эталонного реестра содержится <b>125392</b> объекта

В НСИ из 972 мероприятий Эталонного реестра <b>176</b> мероприятий содержат информацию о <b>125392</b> объектах.
В выгрузке из ГИИС "Электронный бюджет" <b>3523</b> объекта привязаны к <b>132</b> результатам федеральных проектов.

Ни по каким id у нас совпадений в выгрузках нет, придется соединять полученные таблицы по по наименованию мероприятия (результата).

Получилось пересечение по <b>24</b> мероприятиям. Выявлена разница по количеству объектов в Эталонном реестре и ГИИС "Электронный бюджет"
Таблица с результатом выгружена в Excel.
    
Общий вывод. Для более точного анализа необходим общий id для совмещения мероприятий из НСИ и ГИИС "Электронный бюджет".
</div>